design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,480 @@
+    of the `Package` then no further action is needed, assuming the current
+    `PackageRevision` is healthy. This will only occur if the image tag changed
+    but the contents did not, and any previous install `Job` was manually
+    deleted.
+3. Check to see if a `PackageRevision` exists that has name equivalent to the
+   `sha256` hash of the image and is labeled with the `Package` name.
+   - If one does and it is `Active`, no further operation is needed.
+   - If one does and it is `Inactive`, mark it as `Pending` and continue.
+   - If one does and it is `Pending`, continue.
+   - If one does not, create a new `PackageRevision` (or locate an existing one)
+     with a name that matches the `sha256` of the image in `spec.package`, is
+     owned by the `Package`, has a `spec.state` of `Pending`, and is labeled
+     with the name of the `Package`.
+    - Note: If there are other `PackageRevision`s owned by this `Package` that
+      have state `Pending` they should be marked as `Inactive`. This could occur
+      if a previous upgrade was deemed unsafe or a `Package` was updated
+      multiple times in quick succession.
+4. Compare the `Pending` `PackageRevision` to the `Active` `PackageRevision` and
+   determine if it is safe to upgrade. If it is not safe to upgrade, do not
+   proceed. User will be required to change `spec.package` back to previous
+   version if they want to cancel upgrade. Until doing so, the previous
+   `PackageRevision` will continue to be `Active`.
+  - Safe upgrade will initially only be a check that support for CRDs for which
+    instances currently exist or are depended on by other `Package`s will not be
+    dropped, but can be expanded to be more sophisticated in the future.
+5. If upgrade is deemed safe, the `Active` `PackageRevision` should be marked as
+   `Stopping`, and controller should wait until the controller deployed by that
+   `PackageRevision` has been stopped (i.e. `Deployment` no longer exists).
+6. When the `Stopping` `PackageRevision` has finished cleaning up its
+   controller, it should be marked as `Inactive` and the new `PackageRevision`
+   should be marked as `Active`. This will cause it to install all of its
+   objects (updating any existing ones), and start its own controller (i.e.
+   create a `Deployment`).
+7. Wait for `PackageRevision` to become `Ready` (this will occur after
+   successful install of all objects included in the `PackageRevision`).
+8. Set the `status.currentHash` of the `Package` to match the `sha256` hash of
+   the package image and `status.currentImage` to match the package image name.
+
+
+*Deleting a `Package`*
+
+1. Observe delete of a `Package`.
+2. Get the `Active` `PackageRevision` and check that there are no other `Active`
+   package revisions (for other `Package`s) that depend on CRDs that are managed
+   by this `Package` and mark as `Stopping`.
+3. When the `Stopping` `PackageRevision` has finished cleaning up its controller
```

Comment:
@muvaf the `PackageRevision` controller would not have finalizers on the CRDs in this model because the CRDs may persist from one revision to the next. We would likely want the finalizer to be handled by the `Package` controller, which brings up the question of maybe just letting the `Package` controller manage the create / update / delete of the CRDs as opposed to deferring that responsibility to the `PackageRevision`. Then the `PackageRevision` controller would essentially just be responsible for adding ownership labels to CRDs and creating / managing the `Deployment`.

Reply:
Actually, I was meaning the finalizers on the instance of CRD. For example, if I have `GKECluster` CR up and running and the controller stopped, then CRD deletion will take forever because there is a finalizer on `GKECluster` that needs to be removed by the controller.

Reply:
I see. This could be solved by first cleaning up all of the CRs, but we may want to put some checks on when we want to do that such that a user doesn't just delete all of their GCP infrastructure for instance by accidentally deleting the `Package`. Do you think it is unreasonable to say that we won't uninstall a `Package` until all of its dependencies (other packages + instances of its CRDs) have been deleted / uninstalled?